File: list/Solution61.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution61 {
    public ListNode rotateRight(ListNode head, int k) {
    	if(head == null || head.next == null || k == 0){
    		return head;
    	}
    	int size = size(head);
    	for(int i = 0; i < k % size; i++){
    		head = rotateonce(head);
    	}
    	return head;
    }

    private ListNode rotateonce(ListNode head){
    	ListNode point = head;
    	while(point.next.next != null){
    		point = point.next;
    	}
    	point.next.next = head;
    	ListNode result = point.next;
    	point.next = null;
    	return result;
    }

    private int size(ListNode head){
    	ListNode point = head;
    	int size = 0;
    	while(point != null){
    		size += 1;
    		point = point.next;
    	}
    	return size;
    }
}

File: C++/Solution61.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
using namespace std;

 struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
  };
 

class Solution61 {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(!head){
            return head;
        }
        int size = 1;
        ListNode* sentinel = new ListNode(0, head);
        ListNode* ptr = head;
        while(ptr -> next){
            ptr = ptr -> next;
            size++;
        }
        k = k % size;
        if(k == 0){
            return head;
        }
        ListNode* pre = sentinel;
        ListNode* ptr1 = pre -> next;
        ListNode* ptr2 = ptr1;
        for(int i = 0; i < k; i++){
            ptr2 = ptr2 -> next;
        }
        while(ptr2){
            pre = pre -> next;
            ptr1 = ptr1 -> next;
            ptr2 = ptr2 -> next;
        }
        pre -> next = nullptr;
        ptr -> next = head;
        return ptr1;
    }
};